### ISSUE TITLE
RobustDpatch doesn't work

### ISSUE THREAD

**Issue by YXU300** [Trusted-AI/adversarial-robustness-toolbox#2507]
`attack = RobustDPatch(
    estimator=detector,
    patch_shape=(3, 100, 100),
    patch_location=(0,0),
    crop_range=[0, 0],
    brightness_range=[1.0, 1.0],
    rotation_weights=[1, 0, 0, 0],
    sample_size=1,
    learning_rate=1.99,
    max_iter=200,
    batch_size=1,
    verbose=True,
    targeted=True
)

loss_history = []

for images, image_names in get_images(data_dir):
    dets = detector.predict(images)
    y = []
    for j in range(len(dets)):

        highest_box = filter_highest_score_box([dets[j]])
        y.append(highest_box)

    patch = attack.generate(x=images,y=y)
    patched_images = attack.apply_patch(images)

    loss_components, _ = detector._get_losses(patched_images, dets)
    losses = {}
    for loss in loss_components.keys():
        if 'loss' in loss:
            losses[loss] = loss_components[loss].item()
    loss_history.append(losses)
    print(f"Current Losses: {losses}")

    for i in range(len(patched_images)):
        adversarial_image_pgd_bgr = cv2.cvtColor(
            patched_images[i].transpose((1, 2, 0)).astype(np.uint8),
            cv2.COLOR_RGB2BGR
        )
        cv2.imwrite(os.path.join(save_dir, image_names[i]), adversarial_image_pgd_bgr)

print(f"Loss history over all batches: {loss_history}")
`
I tried to attack yolov5s, but it didn't work. The dataset is coco2017.
![13321728791086_ pic](https://github.com/user-attachments/assets/ff293e45-5232-4f7d-8208-17c4350fc899)

**Comment by beat-buesser** [comment]
Hi @YXU300 Thank you very much for your interest in ART! How did you define `detector` in your script above?

**Comment by YXU300** [comment]
import os

import cv2
import numpy as np
from PIL import Image
import torch
from matplotlib import pyplot as plt
from art.attacks.evasion import RobustDPatch
import yolov5
from yolov5.utils.loss import ComputeLoss
from art.estimators.object_detection import PyTorchYolo
import torchvision.transforms as transforms
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader

COCO_INSTANCE_CATEGORY_NAMES = ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
        'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
        'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
        'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
        'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
        'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard',
        'cell phone', 'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors',
        'teddy bear', 'hair drier', 'toothbrush']

def extract_predictions(predictions_, conf_thresh):

    predictions_class = [COCO_INSTANCE_CATEGORY_NAMES[i] for i in list(predictions_["labels"])]
    if len(predictions_class) < 1:
        return [], [], []
    predictions_boxes = [[(i[0], i[1]), (i[2], i[3])] for i in list(predictions_["boxes"])]
    predictions_score = list(predictions_["scores"])
    threshold = conf_thresh
    predictions_t = [predictions_score.index(x) for x in predictions_score if x > threshold]
    if len(predictions_t) > 0:
        predictions_t = predictions_t
    else:
        return [], [], []
    # predictions in score order
    predictions_boxes = [predictions_boxes[i] for i in predictions_t]
    predictions_class = [predictions_class[i] for i in predictions_t]
    predictions_scores = [predictions_score[i] for i in predictions_t]
    return predictions_class, predictions_boxes, predictions_scores




def filter_highest_score_box(predictions):
    highest_score = -1
    highest_score_box = None
    highest_score_label = None

    for pred in predictions:
        scores = pred["scores"]
        if len(scores) == 0:
            continue

        max_score_index = np.argmax(scores)
        max_score = scores[max_score_index]

        if max_score > highest_score:
            highest_score = max_score
            highest_score_box = pred["boxes"][max_score_index]
            highest_score_label = pred["labels"][max_score_index]

    if highest_score_box is not None:
        return {
            "boxes": np.array([highest_score_box]),
            "scores": np.array([highest_score]),
            "labels": np.array([highest_score_label])
        }

    return {"boxes": np.array([]), "scores": np.array([]), "labels": np.array([])}




class Yolo(torch.nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.model.hyp = {'box': 0.1, 'obj': 0.1, 'cls': 0.1, 'anchor_t': 1.0, 'cls_pw': 1.0, 'obj_pw': 1.0, 'fl_gamma': 0.0}
        self.compute_loss = ComputeLoss(self.model.model.model)

    def forward(self, x, targets=None):
        if self.training:
            outputs = self.model.model.model(x)
            loss, loss_items = self.compute_loss(outputs, targets)
            loss_components_dict = {"loss_total": loss, 'loss_box': loss_items[0], 'loss_obj': loss_items[1], 'loss_cls': loss_items[2]}
            return loss_components_dict
        else:
            return self.model(x)

model = torch.hub.load('ultralytics/yolov5', 'yolov5s')
model = Yolo(model)
detector = PyTorchYolo(model=model,
                       device_type='cuda',
                       input_shape=(3, 256, 256),
                       clip_values=(0, 200),
                       attack_losses=("loss_total", "loss_cls", "loss_box", "loss_obj"))

image_filenames = [f for f in os.listdir(data_dir) if f.endswith(('.jpg', '.png'))]
NUMBER_CHANNELS = 3
INPUT_SHAPE = (NUMBER_CHANNELS, 256, 256)
transform = transforms.Compose([
    transforms.Resize(INPUT_SHAPE[1], interpolation=transforms.InterpolationMode.BICUBIC),
    transforms.CenterCrop(INPUT_SHAPE[1]),
    transforms.ToTensor()
])
def get_images(image_dir, batch_size=1):
    image_files = [f for f in os.listdir(image_dir) if f.endswith('.jpg') or f.endswith('.png')]
    for i in range(0, len(image_files), batch_size):
        images = []
        image_names = []
        for image_file in image_files[i:i+batch_size]:
            im = Image.open(os.path.join(image_dir, image_file))
            im = transform(im).numpy()
            images.append(im)
            image_names.append(image_file)
        yield np.array(images)*255, image_names


attack = RobustDPatch(
    estimator=detector,
    patch_shape=(3, 100, 100),
    patch_location=(0,0),
    crop_range=[0, 0],
    brightness_range=[1.0, 1.0],
    rotation_weights=[1, 0, 0, 0],
    sample_size=1,
    learning_rate=1.99,
    max_iter=200,
    batch_size=5,
    verbose=True,
    targeted=False
)

loss_history = []

for images, image_names in get_images(data_dir):
    dets = detector.predict(images)
    y = []
    for j in range(len(dets)):

        highest_box = filter_highest_score_box([dets[j]])
        y.append(highest_box)

    patch = attack.generate(x=images)
    patched_images = attack.apply_patch(images)

    loss_components, _ = detector._get_losses(patched_images, dets)
    losses = {}
    for loss in loss_components.keys():
        if 'loss' in loss:
            losses[loss] = loss_components[loss].item()
    loss_history.append(losses)
    print(f"Current Losses: {losses}")

    for i in range(len(patched_images)):
        adversarial_image_pgd_bgr = cv2.cvtColor(
            patched_images[i].transpose((1, 2, 0)).astype(np.uint8),
            cv2.COLOR_RGB2BGR
        )
        cv2.imwrite(os.path.join(save_dir, image_names[i]), adversarial_image_pgd_bgr)

print(f"Loss history over all batches: {loss_history}")

**Comment by YXU300** [comment]
Thanks for your reply, here is my code.

**Comment by YXU300** [comment]
I've also tried to run the example code and it also has no work.

**Comment by beat-buesser** [comment]
Closing because of inactivity, please reopen if still current.
